test/worker.test.js

```diff
@@ -0,0 +1,32 @@
+import { describe, it, expect, vi } from 'vitest';
+import worker from '../bin/worker.js';
+import { reportMetric } from '../lib/influx.js';
+
+
+
+describe('worker.fetch', () => {
+  it('should fetch and report metrics correctly', async () => {
+    const request = {
+      url: 'https://example.com/path?api-key=test-key',
+      method: 'GET'
+    };
+    const env = {
+      REQUEST_URL: 'https://proxy.example.com'
+    };
+    const ctx = {
+      waitUntil: vi.fn()
+    };
+
+    global.fetch = vi.fn().mockResolvedValue(new Response(null, { status: 200 }));
+    const response = await worker.fetch(request, env, ctx);
+
+    expect(global.fetch).toHaveBeenCalledWith(
+      expect.objectContaining({
+        url: 'https://example.com/path?api-key=test-key',
+        method: 'GET'
+      })
+    );
+    expect(ctx.waitUntil).toHaveBeenCalledWith(reportMetric(request, response, env));
```

Comment:
I am not familiar with vitest and their implementation of function mocks. Could you please explain how do you expect this line to work?

In particular: `reportMetric(request, response, env)` returns a promise resolving to the result of calling `fetch()` to write to InfluxDB. Comparing the promise created while running `worker.fetch` with the promise created by calling `reportMetric` directly seems suspicious to me.

I think a better way how to go about this is to inject `reportMetric` implementation into the worker code, so that we can write the mock at the higher level of abstraction.


```js
// in bin/worker.js
import { reportMetric } from '../lib/influx.js';

export default {
  async fetch(request, env, ctx, { reportMetric } = { reportMetric }) {
    // ...
  }
}

// in the test

reportMetric = vi.fn().mockResolvedValue(new Promise(new Response()))
const response = await worker.fetch(request, env, ctx, { reportMetric })
expect(reportMetric). toHaveBeenCalledWith(request, env)
```

Thoughts?

Reply:
I do like your suggestion, although I have tried to test if `ctx.waitUntil` has been called with `Promise<Response>` directly and that seems to not work. 

I am going to insert your suggestion, but I think it would also be valuable to test for `ctx.waitUntil` call as that ensures that reporting of metrics happens after returning response and that it happens in non-blocking way.